components/edit-func.js

```diff
@@ -55,6 +79,23 @@ function JsonEditor({ text, onChange, enabled, errors = [] }) {
     }
   }, [darkMode.enabled]);
 
+  useEffect(() => {
+    const newErrors = jsonErrors(text);
+    // If we have a new error,
+    // or if we previously had an error but now have [] then set Errors
+    if (
+      (newErrors.length > 0 && !_.isEqual(newErrors, errors)) ||
+      (newErrors.length === 0 && errors.length > 0)
+    ) {
+      if (onError !== null) {
+        onError(newErrors);
+      }
+      setErrors(newErrors);
+    }
+  }, [text, errors, onError]);
```

Comment:
Yeah as this evolved, I think my separation was that `edit-func` is for the editor itself, and nothing else. But `edit` is for handling Saving json and being aware that this is an IGVF object. This is in contrast to `add` which knows that the editor is for Adding a new object and everything to do with that.

It's also true that the distinction between those could be coalesced at some point because the only real difference is the backend routes we're using and how we generate the initial contents of the editor.

Reply:
The other thing is that I was having an infinite render loop error when I did not have this in a useEffect. I previously had an `internalOnChange` which called the passed in the prop `onChange` and then did the error updates. But maybe I was doing something weird? I tried a few different things and the internet had said that I should put it in the useEffect

Reply:
Infinite loops are more common with `useEffect`, because the callback changes a state, which causes a `useEffect` dependency to change, which calls the `useEffect` callback, which changes the state again, which causes a dependency to change, etc.

The change handler should only get called with a change. So then it should (and does) change the state that holds the editor contents, and checks for bad input, which should disable the Save button, and that should be it. The onChange callback shouldn’t be called again until the user types something else.